ballerina/websocket_service.bal

```diff
@@ -24,70 +25,111 @@ isolated service class WsService {
+                    closeConnection(caller, 4429, "Too many initialisation requests");
+                    return;
+                }
+                self.initiatedConnection = true;
+            }
+            json wsResponse = {"type": WS_ACK};
+            check caller->writeTextMessage(wsResponse.toJsonString());
+        } else if wsType == WS_SUBSCRIBE || wsType == WS_START || !self.customHeaders.hasKey(WS_SUB_PROTOCOL) {
+            lock {
+                if self.customHeaders.hasKey(WS_SUB_PROTOCOL) {
+                    if !self.initiatedConnection {
+                        closeConnection(caller, 4401, "Unauthorized");
+                        return;
+                    }
+                    if self.activeConnections.indexOf(connectionId) !is () {
+                        closeConnection(caller, 4409, string `Subscriber for ${connectionId} already exists`);
+                        return;
+                    }
+                    string[] connections = [connectionId];
+                    foreach string i in self.activeConnections {
+                        connections.push(i);
                     }
-                    next = sourceStream.iterator().next();
+                    self.activeConnections = connections.cloneReadOnly();
                 }
+            }
+            parser:OperationNode|json node = validateSubscriptionPayload(wsPayload, self.engine);
+            if node is parser:OperationNode {
+                check executeOperation(self.engine, self.context, self.schema, self.customHeaders, caller,
+                                       connectionId, node);
             } else {
-                check caller->writeTextMessage(sourceStream.toJsonString());
+                check sendWebSocketResponse(caller, self.customHeaders, WS_ERROR, node, connectionId);
+                closeConnection(caller);
+            }
+        } else if (wsType.equalsIgnoreCaseAscii(WS_STOP) || wsType.equalsIgnoreCaseAscii(WS_COMPLETE)) {
+            lock {
+                string[] connections = [];
+                foreach string i in self.activeConnections {
+                    connections.push(i);
+                }
+                int? index = array:indexOf(connections, connectionId);
+                if index is int {
+                    _ = connections.remove(index);
+                }
+                self.activeConnections = connections.cloneReadOnly();
+                self.initiatedConnection = false;
             }
+            check sendWebSocketResponse(caller, self.customHeaders, WS_COMPLETE, null, connectionId);
+            closeConnection(caller);
+        } else if wsType.equalsIgnoreCaseAscii(WS_PING) {
+            json wsMessage = {"type": WS_PONG};
+            check caller->writeTextMessage(wsMessage.toString());
+        } else if wsType.equalsIgnoreCaseAscii(WS_PONG) {
+            json wsMessage = {"type": WS_PING};
+            check caller->writeTextMessage(wsMessage.toString());
```

Comment:
It seems we can handle these two cases with [remote methods](https://github.com/ballerina-platform/module-ballerina-websocket/blob/main/docs/spec/spec.md#onping-and-onpong) associated with the WebSocket service. @Nuvindu can you check this as well?

Reply:
As I understand, `ping` and `pong` specified in the [graphql-ws spec](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#ping) are the types wrapped in text messages. So they should be handled in `onTextMessage()` function. Is that the correct approach?

Reply:
In that case, this is fine.